### DISCUSSION TITLE
drive a screen with r1,g1,b1,r2,g2,b2,r3,g3,b3

### DISCUSSION THREAD

**Discussion by yuhan1965** [mrcodetastic/ESP32-HUB75-MatrixPanel-DMA#598] [Q&A] [answered]
I am using a 64X32 16-scan screen, and everything is normal. Now I have a 96x48 screen, which is also 16-scan, but it has more color PINs R3, G3, B3. How can I modify the program to make it work properly? I have made the following changes, but they are not successful. Could you please give me some guidance? Thank you!
![layout](https://github.com/mrfaptastic/ESP32-HUB75-MatrixPanel-DMA/assets/136178331/1d559f88-7b66-4f6f-9638-56403d8a048f)
------------------------------------------------
#define MATRIX_ROWS_IN_PARALLEL 3
-------------------------------------------------
#define BITS_RGB3_OFFSET 13 // Start point of RGB_X2 bits
#define BIT_R3 (1 << 13)
#define BIT_G3 (1 << 14)
#define BIT_B3 (1 << 15)

// BitMasks are pre-computed based on the above #define's for performance.
#define BITMASK_RGB1_CLEAR (0b1111111111111000)  // inverted bitmask for R1G1B1 bit in pixel vector
#define BITMASK_RGB2_CLEAR (0b1111111111000111)  // inverted bitmask for R2G2B2 bit in pixel vector
#define BITMASK_RGB3_CLEAR (0b0001111111111111)  // inverted bitmask for R3G3B3 bit in pixel vector by SSM
//#define BITMASK_RGB12_CLEAR (0b1111111111000000) // inverted bitmask for R1G1B1R2G2B2 bit in pixel vector
#define BITMASK_RGB123_CLEAR (0b0001111111000000) // inverted bitmask for R1G1B1R2G2B2R3G3B3 bit in pixel vector
#define BITMASK_CTRL_CLEAR (0b1110000000111111)  // inverted bitmask for control bits ABCDE,LAT,OE in pixel vector
#define BITMASK_OE_CLEAR (0b1111111101111111)    // inverted bitmask for control bit OE in pixel vector
-------------------------------------------------------------------------

  uint16_t _colourbitclear = BITMASK_RGB1_CLEAR, _colourbitoffset = 0;

  if (y_coord >= ROWS_PER_FRAME)
  { // if we are drawing to the bottom part of the panel
    _colourbitoffset = BITS_RGB2_OFFSET;
    _colourbitclear = BITMASK_RGB2_CLEAR;
    y_coord -= ROWS_PER_FRAME;
  }
//my adding ...
  if (y_coord >= 2*ROWS_PER_FRAME)
  { // if we are drawing to the bottom part of the panel
    _colourbitoffset = BITS_RGB3_OFFSET;
    _colourbitclear = BITMASK_RGB3_CLEAR;
    y_coord -= 2*ROWS_PER_FRAME;
  }

**Comment by mrcodetastic** [comment]
Nearly there. The bit you're missing is the bit where you enable these additional bits to be output electrically into the real world. The hardware doesn't know where to send these specific bits otherwise. 

In `ESP32-HUB75-MatrixPanel-I2S-DMA.cpp` you need to map the bits to a GPIO
```
void MatrixPanel_I2S_DMA::configureDMA(const HUB75_I2S_CFG &_cfg)
{

<snipped>

  bus_cfg.pin_d13 = -1;
  bus_cfg.pin_d14 = -1;
  bus_cfg.pin_d15 = -1;


```

**Reply by yuhan1965** [reply]
Thank you very much for your good prompt. 
In fact, I have already completed the definition and output of the GPIO s, and the program has been edited and passed. 
However, I do not fully understand the code listed above, and not sure if any other parts of the sketch need to be modified.

**Reply by mrcodetastic** [reply]
Sorry.  I don't understand what you are saying.  Are you using a translator?

**Reply by yuhan1965** [reply, marked as answer]
I put the adding sketch before the IF sentence, like below, it works.

//my adding ...
if (y_coord >= 2*ROWS_PER_FRAME)
{ // if we are drawing to the bottom part of the panel
_colourbitoffset = BITS_RGB3_OFFSET;
_colourbitclear = BITMASK_RGB3_CLEAR;
y_coord -= 2*ROWS_PER_FRAME;
}

if (y_coord >= ROWS_PER_FRAME)
{ // if we are drawing to the bottom part of the panel
_colourbitoffset = BITS_RGB2_OFFSET;
_colourbitclear = BITMASK_RGB2_CLEAR;
y_coord -= ROWS_PER_FRAME;
}

**Reply by mrcodetastic** [reply]
So you got it to work without issue?

**Reply by yuhan-1965** [reply]
Yes, thank you for affirming my idea at the beginning, and it is now working perfectly!
![5ad2cb743887f1bd221e056e8e78499](https://github.com/mrfaptastic/ESP32-HUB75-MatrixPanel-DMA/assets/153521271/7a8c4a6d-91e9-460b-a3b3-1727f718f74b)

**Reply by harissutanrafiq** [reply]
> Yes, thank you for affirming my idea at the beginning, and it is now working perfectly! ![5ad2cb743887f1bd221e056e8e78499](https://private-user-images.githubusercontent.com/153521271/316189941-7a8c4a6d-91e9-460b-a3b3-1727f718f74b.jpg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjE5MDczMDUsIm5iZiI6MTcyMTkwNzAwNSwicGF0aCI6Ii8xNTM1MjEyNzEvMzE2MTg5OTQxLTdhOGM0YTZkLTkxZTktNDYwYi1hM2IzLTE3MjdmNzE4Zjc0Yi5qcGc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNzI1JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDcyNVQxMTMwMDVaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0yMDdiMzk4OGNlYzJjODI4MWY2ZjhmYjNhODEzZGZmYzk1MGUwMTM4N2E3OTgyZjM4YTc1ODg5ZDUzNjlhM2NlJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.4JhzCHdkpGwuX7UC4bkB09vB9BRpClUc23cqBfsJ3PA)

Hey can you share library you have modified ? I have same panel
